### PR TITLE
MAGN-7851, MAGN-8334

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose

This PR addresses [MAGN-7851](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7851) and [MAGN-8334](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8334).

For MAGN-7851, we've added a new method to calculate the near and far clip plane distances to accurately encapsulate the scene, `HelixWatch3DViewModel.ComputeClipPlaneDistances(...)`. Previously, when a object far away was added to Dynamo, the near clip plane would adjust to a high value, and the near geometries would get clipped.

For MAGN-8334, it was found while fixing MAGN-7851, that the HelixToolkit was not properly calculating the bounds for `GeometryModel3D` objects whose geometry was updated. This was most obvious when placing a series of points in Dynamo. The `Bounds` property for the `PointGeometryModel3D` object would only return the bounds of the first point in the series. What's worse is that, because it's a point, the bounds would return a zero-size bounding box. An extension method, `GeometryModel3D.Bounds()` has been added which correctly calculates the bounds of point objects AND adds sizes the bounds so that the point is in the middle and the bounds have a size of 5 units. 

### Bonus Feature
FYI @Racel @lillismith 
When fixing MAGN-7851 it was impossible to test whether the clip planes were working, because we could not zoom to fit around and object way out in space. This PR also adds overload functionality to the "Zoom to Fit" functionality. The Zoom to Fit modes are now as follows:
- Zoom to Fit with nothing selected. Zooms to fit all geometry.
- Zoom to Fit with nodes selected. Zooms to fit around geometry from the selected nodes.
The zoom to fit operation already re-centered the point of rotation, so you are now able to zoom to very remote objects and tumble them view around their center.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
  - No new user-facing strings added.

### Reviewers

@ramramps

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Watch3D/DefaultWatch3DViewModel.cs [hunk 1]
```diff
@@ -196,7 +198,7 @@ public DefaultWatch3DViewModel(Watch3DViewModelStartupParams parameters)
             TogglePanCommand = new DelegateCommand(TogglePan, CanTogglePan);
             ToggleOrbitCommand = new DelegateCommand(ToggleOrbit, CanToggleOrbit);
             ToggleCanNavigateBackgroundCommand = new DelegateCommand(ToggleCanNavigateBackground, CanToggleCanNavigateBackground);
-
+            ZoomToFitCommand = new DelegateCommand(ZoomToFit, CanZoomToFit);
```
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/Watch3D/DefaultWatch3DViewModel.cs]: ZoomToFit is now a command on the view model instead of a click event handler in the view's code behind.

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs [hunk 2]
```diff
@@ -59,6 +61,10 @@ private void UnregisterEventHandlers()
             if (ViewModel == null) return;
 
             ViewModel.RequestAttachToScene -= ViewModelRequestAttachToSceneHandler;
+            ViewModel.RequestCreateModels -= RequestCreateModelsHandler;
```
Comment (ikeough) [inline src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs]: These event handlers were not previously unregistered.

**Comment by Racel** [comment]
@ikeough - very cool! can't wait to try it out.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 3]
```diff
@@ -151,6 +156,15 @@ public void OnRequestCreateModels(IEnumerable<IRenderPackage> packages)
             }
         }
 
+        public event Action<BoundingBox> RequestZoomToFit;
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: Comment  please.

**Comment by jnealb** [comment]
@ikeough Very interesting.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 4]
```diff
@@ -1240,15 +1270,111 @@ private static MeshGeometry3D DrawTestMesh()
...
+
+            // See http://mathworld.wolfram.com/Point-PlaneDistance.html
+            // The plane distance formula will return positive values for points on the same side of the plane
+            // as the plane's normal, and negative values for points on the opposite side of the plane. 
+
+            var distances = bounds.SelectMany(b => b.GetCorners()).
+                Select(c => c.DistanceToPlane(cameraPosition, cameraLook.Normalized())).
+                ToList();
+
+            distances.Sort();
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: this should be inside the if loop.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 5]
```diff
@@ -1240,15 +1270,111 @@ private static MeshGeometry3D DrawTestMesh()
...
+
+            if (distances.Any())
+            {
+                // All behind
+                // Set the near and far clip to their defaults
+                // because nothing is in front of the camera.
+                if (distances.All(d => d < 0))
+                {
+                    //Debug.WriteLine("All points behind.");
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: Please remove the commented Debug.WriteLine lines.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 6]
```diff
@@ -1240,15 +1270,111 @@ private static MeshGeometry3D DrawTestMesh()
...
+                    near = defaultNearClipDistance;
+                    far = defaultFarClipDistance;
+                }
+                // All in front
+                // Set the near clip plane to some fraction of the 
+                // of the distance to the first point.
+                else if (distances.All(d => d > 0))
+                {
+                    //Debug.WriteLine("All points in front.");
+                    near = Math.Max(defaultNearClipDistance, distances.First(d => d > 0) * nearPlaneDistanceFactor);
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: I think this should be just distances.First() *  nearPlaneDistanceFactor. Because, this loop executes for all distance > 0, that case first one will be always > 0. 
If distances.First(d => d > 0) is required like this, (I think) else if and else loop executes the same. Do we really need the else part?   suppose you have the distance as -1, 0 ,1,2,3 (after sorting), both these line executes have distances.first as 1, and last as 3.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 7]
```diff
@@ -1240,15 +1270,111 @@ private static MeshGeometry3D DrawTestMesh()
...
+            
+            var bounds =
+                geometry.Where(i => i is GeometryModel3D)
+                    .Cast<GeometryModel3D>().Select(g=>g.Bounds());
+
+            // See http://mathworld.wolfram.com/Point-PlaneDistance.html
+            // The plane distance formula will return positive values for points on the same side of the plane
+            // as the plane's normal, and negative values for points on the opposite side of the plane. 
+
+            var distances = bounds.SelectMany(b => b.GetCorners()).
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: Do we need to check bounds.any here?

**Comment by jnealb** [comment]
@ikeough @RodRecker  Can you please comment on the testing effort required, especially as regards to the 9.0 release? Is this safe for 9.0, or should we defer?

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs (hunk 6 above)
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: You are correct @ramramps. We can merge the last two cases. I hadn't realized late at night that they were the same. The `d=> d>0` then has to stay because the `distances` collection may have values `<0`.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs (hunk 7 above)
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: Actually, we should check if `geometry` is empty. Then fall back to defaults.

**Comment by ikeough** [comment]
@jnealb The best I can say is that it fixes the problem indicated in the bug, and adds tests around this functionality that were not there before. But, there are a lot of strange edge conditions that might be hit with large Revit models.
